File: implementations/ts/pcf-sig/src/manifest.ts
/**
 * The Manifest and Signed Entry stored in a `PCFSIG_SIG` partition
 * (spec Section 7).
 *
 * The Manifest is the byte sequence that is hashed and signed. Its length is
 * deterministic from `signedCount`:
 *   `MANIFEST_PREFIX_SIZE + SIGNED_ENTRY_SIZE * signedCount`.
 */

import {
  HASH_FIELD_SIZE,
  HashAlgo,
  TYPE_RESERVED,
  UID_SIZE,
  hashAlgoFromId,
  hashAlgoId,
} from "@kduma-oss/pcf";

import {
  SigAlgo,
  requiredManifestHash,
  sigAlgoFromId,
  sigAlgoId,
} from "./algo.js";
import {
  FINGERPRINT_SIZE,
  MANIFEST_PREFIX_SIZE,
  PROFILE_VERSION_MAJOR,
  PROFILE_VERSION_MINOR,
  SIG_MAGIC,
  SIGNED_ENTRY_SIZE,
} from "./consts.js";
import { PcfSigError } from "./errors.js";

/** Whether a PCF hash algorithm id is cryptographic (spec Section 9). */
export function isCryptoHash(algo: HashAlgo): boolean {
  return (
    algo === HashAlgo.Sha256 ||
    algo === HashAlgo.Sha512 ||
    algo === HashAlgo.Blake3
  );
}

/** One Signed Entry inside a Manifest (spec Section 7.2). */
export interface SignedEntry {
  /** PCF uid of the covered partition (verbatim). */
  uid: Uint8Array;
  /** PCF type of the covered partition (verbatim). */
  partitionType: number;
  /** PCF label of the covered partition (verbatim 32-byte field). */
  label: Uint8Array;
  /** PCF `used_bytes` of the covered partition. */
  usedBytes: bigint;
  /** PCF `data_hash_algo_id`. MUST be cryptographic in v1.0 (16/17/18). */
  dataHashAlgo: HashAlgo;
  /** PCF `data_hash` field bytes (verbatim 64-byte field). */
  dataHash: Uint8Array;
}

/** Serialise a Signed Entry to its on-disk 218-byte layout (spec Section 7.2). */
export function signedEntryToBytes(e: SignedEntry): Uint8Array {
  const b = new Uint8Array(SIGNED_ENTRY_SIZE);
  const view = new DataView(b.buffer);
  b.set(e.uid, 0);
  view.setUint32(16, e.partitionType >>> 0, true);
  b.set(e.label, 20);
  view.setBigUint64(52, e.usedBytes, true);
  b[60] = hashAlgoId(e.dataHashAlgo);
  // b[61] reserved = 0
  b.set(e.dataHash, 62);
  // b[126..218] reserved = 0
  return b;
}

/**
 * Parse a Signed Entry from its on-disk 218-byte layout. Validates the
 * reserved spans, the cryptographic-hash constraint (Section 9), and the PCF
 * reserved-value guards (Section 11, V7).
 */
export function signedEntryFromBytes(b: Uint8Array): SignedEntry {
  if (b.length !== SIGNED_ENTRY_SIZE) {
    throw PcfSigError.malformedSignaturePartition();
  }
  if (b[61] !== 0) {
    throw PcfSigError.nonZeroEntryReserved();
  }
  for (let i = 126; i < 218; i++) {
    if (b[i] !== 0) {
      throw PcfSigError.nonZeroEntryReserved();
    }
  }
  const uid = b.slice(0, UID_SIZE);
  if (uid.every((x) => x === 0)) {
    throw PcfSigError.entryNilUid();
  }
  const view = new DataView(b.buffer, b.byteOffset, b.byteLength);
  const partitionType = view.getUint32(16, true);
  if (partitionType === TYPE_RESERVED) {
    throw PcfSigError.entryReservedType();
  }
  const label = b.slice(20, 52);
  const usedBytes = view.getBigUint64(52, true);
  const dataHashAlgo = hashAlgoFromId(b[60]!);
  if (!isCryptoHash(dataHashAlgo)) {
    throw PcfSigError.nonCryptoEntryHash(b[60]!);
  }
  const dataHash = b.slice(62, 62 + HASH_FIELD_SIZE);
  return {
    uid,
    partitionType,
    label,
    usedBytes,
    dataHashAlgo,
    dataHash,
  };
}

/** A parsed Manifest (spec Section 7.1). */
export interface Manifest {
  /** `manifest_version_major`. */
  versionMajor: number;
  /** `manifest_version_minor`. */
  versionMinor: number;
  /** `sig_algo_id`. */
  sigAlgo: SigAlgo;
  /** `manifest_hash_algo_id`. MUST be cryptographic (16/17/18). */
  manifestHashAlgo: HashAlgo;
  /** Reserved `flags` field; v1.0 MUST be 0. */
  flags: number;
  /** Signer key fingerprint. */
  signerKeyFingerprint: Uint8Array;
  /** `signed_at_unix_seconds` (i64). */
  signedAtUnixSeconds: bigint;
  /** `signed_entries`, packed in writer-chosen order. */
  signedEntries: SignedEntry[];
}

/** Construct a Manifest from its component parts. */
export function makeManifest(
  sigAlgo: SigAlgo,
  manifestHashAlgo: HashAlgo,
  signerKeyFingerprint: Uint8Array,
  signedAtUnixSeconds: bigint,
  signedEntries: SignedEntry[],
): Manifest {
  return {
    versionMajor: PROFILE_VERSION_MAJOR,
    versionMinor: PROFILE_VERSION_MINOR,
    sigAlgo,
    manifestHashAlgo,
    flags: 0,
    signerKeyFingerprint: new Uint8Array(signerKeyFingerprint),
    signedAtUnixSeconds,
    signedEntries,
  };
}

/** Serialised length in bytes. */
export function manifestByteLen(m: Manifest): number {
  return MANIFEST_PREFIX_SIZE + SIGNED_ENTRY_SIZE * m.signedEntries.length;
}

/** Serialise a Manifest to the on-disk byte layout (spec Section 7.1). */
export function manifestToBytes(m: Manifest): Uint8Array {
  const out = new Uint8Array(manifestByteLen(m));
  const view = new DataView(out.buffer);
  out.set(SIG_MAGIC, 0);
  view.setUint16(8, m.versionMajor, true);
  view.setUint16(10, m.versionMinor, true);
  out[12] = sigAlgoId(m.sigAlgo);
  out[13] = hashAlgoId(m.manifestHashAlgo);
  view.setUint16(14, m.flags, true);
  out.set(m.signerKeyFingerprint, 16);
  view.setBigInt64(48, m.signedAtUnixSeconds, true);
  view.setUint32(56, m.signedEntries.length, true);
  for (let i = 0; i < m.signedEntries.length; i++) {
    out.set(
      signedEntryToBytes(m.signedEntries[i]!),
      MANIFEST_PREFIX_SIZE + i * SIGNED_ENTRY_SIZE,
    );
  }
  return out;
}

/**
 * Parse a Manifest from the on-disk byte layout. Validates: magic, major
 * version, algorithm registry membership, hash-algo binding (Section 8),
 * cryptographic hash requirement (Section 9), reserved flags, non-empty
 * signed_count, and per-entry reserved spans (Section 7.2). Does NOT validate
 * duplicate uids or self-reference; the verifier does that with context from
 * the enclosing partition.
 */
export function manifestFromBytes(b: Uint8Array): Manifest {
  if (b.length < MANIFEST_PREFIX_SIZE) {
    throw PcfSigError.malformedSignaturePartition();
  }
  if (!bytesEqual(b.subarray(0, 8), SIG_MAGIC)) {
    throw PcfSigError.badManifestMagic();
  }
  const view = new DataView(b.buffer, b.byteOffset, b.byteLength);
  const versionMajor = view.getUint16(8, true);
  const versionMinor = view.getUint16(10, true);
  if (versionMajor !== PROFILE_VERSION_MAJOR) {
    throw PcfSigError.unsupportedMajor(versionMajor);
  }
  const sigAlgo = sigAlgoFromId(b[12]!);
  const manifestHashId = b[13]!;
  const manifestHashAlgo = hashAlgoFromId(manifestHashId);
  if (!isCryptoHash(manifestHashAlgo)) {
    throw PcfSigError.nonCryptoManifestHash(manifestHashId);
  }
  const required = requiredManifestHash(sigAlgo);
  if (required !== null && required !== manifestHashAlgo) {
    throw PcfSigError.hashAlgoBindingMismatch();
  }
  const flags = view.getUint16(14, true);
  if (flags !== 0) {
    throw PcfSigError.nonZeroFlags();
  }
  const signerKeyFingerprint = b.slice(16, 16 + FINGERPRINT_SIZE);
  const signedAtUnixSeconds = view.getBigInt64(48, true);
  const signedCount = view.getUint32(56, true);
  if (signedCount === 0) {
    throw PcfSigError.emptyManifest();
  }
  const expected = MANIFEST_PREFIX_SIZE + SIGNED_ENTRY_SIZE * signedCount;
  if (b.length < expected) {
    throw PcfSigError.malformedSignaturePartition();
  }
  const signedEntries: SignedEntry[] = [];
  const seen = new Set<string>();
  for (let i = 0; i < signedCount; i++) {
    const off = MANIFEST_PREFIX_SIZE + i * SIGNED_ENTRY_SIZE;
    const e = signedEntryFromBytes(b.slice(off, off + SIGNED_ENTRY_SIZE));
    const key = uidKey(e.uid);
    if (seen.has(key)) {
      throw PcfSigError.duplicateSignedUid();
    }
    seen.add(key);
    signedEntries.push(e);
  }
  return {
    versionMajor,
    versionMinor,
    sigAlgo,
    manifestHashAlgo,
    flags,
    signerKeyFingerprint,
    signedAtUnixSeconds,
    signedEntries,
  };
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function uidKey(uid: Uint8Array): string {
  let s = "";
  for (let i = 0; i < uid.length; i++) {
    s += uid[i]!.toString(16).padStart(2, "0");
  }
  return s;
}

File: implementations/dotnet/pcf-sig/src/Pcf.Sig/Manifest.cs
using System;
using System.Collections.Generic;
using Pcf;

namespace Pcf.Sig;

/// <summary>One Signed Entry inside a Manifest (spec Section 7.2).</summary>
public sealed class SignedEntry
{
    public byte[] Uid { get; set; } = new byte[Pcf.Constants.UidSize];
    public uint PartitionType { get; set; }
    public byte[] Label { get; set; } = new byte[Pcf.Constants.LabelSize];
    public ulong UsedBytes { get; set; }
    public HashAlgo DataHashAlgo { get; set; }
    public byte[] DataHash { get; set; } = new byte[Pcf.Constants.HashFieldSize];

    /// <summary>Serialise to the on-disk 218-byte layout.</summary>
    public byte[] ToBytes()
    {
        var b = new byte[Constants.SignedEntrySize];
        Buffer.BlockCopy(Uid, 0, b, 0, Pcf.Constants.UidSize);
        LittleEndian.WriteU32(b, 16, PartitionType);
        Buffer.BlockCopy(Label, 0, b, 20, Pcf.Constants.LabelSize);
        LittleEndian.WriteU64(b, 52, UsedBytes);
        b[60] = DataHashAlgo.Id();
        // b[61] reserved = 0
        Buffer.BlockCopy(DataHash, 0, b, 62, Pcf.Constants.HashFieldSize);
        // b[126..218] reserved = 0
        return b;
    }

    /// <summary>
    /// Parse from the on-disk 218-byte layout. Validates reserved spans, the
    /// cryptographic-hash constraint (Section 9), and the PCF reserved-value
    /// guards (Section 11, V7).
    /// </summary>
    public static SignedEntry FromBytes(byte[] b)
    {
        if (b == null || b.Length != Constants.SignedEntrySize)
        {
            throw PcfSigException.MalformedSignaturePartition();
        }
        if (b[61] != 0)
        {
            throw PcfSigException.NonZeroEntryReserved();
        }
        for (int i = 126; i < 218; i++)
        {
            if (b[i] != 0)
            {
                throw PcfSigException.NonZeroEntryReserved();
            }
        }
        var uid = new byte[Pcf.Constants.UidSize];
        Buffer.BlockCopy(b, 0, uid, 0, Pcf.Constants.UidSize);
        if (IsAllZero(uid))
        {
            throw PcfSigException.EntryNilUid();
        }
        uint partitionType = LittleEndian.ReadU32(b, 16);
        if (partitionType == Pcf.Constants.TypeReserved)
        {
            throw PcfSigException.EntryReservedType();
        }
        var label = new byte[Pcf.Constants.LabelSize];
        Buffer.BlockCopy(b, 20, label, 0, Pcf.Constants.LabelSize);
        ulong usedBytes = LittleEndian.ReadU64(b, 52);
        var dataHashAlgo = HashAlgoExtensions.FromId(b[60]);
        if (!Manifest.IsCryptoHash(dataHashAlgo))
        {
            throw PcfSigException.NonCryptoEntryHash(b[60]);
        }
        var dataHash = new byte[Pcf.Constants.HashFieldSize];
        Buffer.BlockCopy(b, 62, dataHash, 0, Pcf.Constants.HashFieldSize);
        return new SignedEntry
        {
            Uid = uid,
            PartitionType = partitionType,
            Label = label,
            UsedBytes = usedBytes,
            DataHashAlgo = dataHashAlgo,
            DataHash = dataHash,
        };
    }

    private static bool IsAllZero(byte[] b)
    {
        for (int i = 0; i < b.Length; i++) if (b[i] != 0) return false;
        return true;
    }
}

/// <summary>A parsed Manifest (spec Section 7.1).</summary>
public sealed class Manifest
{
    public ushort VersionMajor { get; set; }
    public ushort VersionMinor { get; set; }
    public SigAlgo SigAlgo { get; set; }
    public HashAlgo ManifestHashAlgo { get; set; }
    public ushort Flags { get; set; }
    public byte[] SignerKeyFingerprint { get; set; } = new byte[Constants.FingerprintSize];
    public long SignedAtUnixSeconds { get; set; }
    public List<SignedEntry> SignedEntries { get; set; } = new();

    /// <summary>Construct a Manifest from its component parts.</summary>
    public static Manifest Make(
        SigAlgo sigAlgo,
        HashAlgo manifestHashAlgo,
        byte[] signerKeyFingerprint,
        long signedAtUnixSeconds,
        List<SignedEntry> signedEntries)
    {
        return new Manifest
        {
            VersionMajor = Constants.ProfileVersionMajor,
            VersionMinor = Constants.ProfileVersionMinor,
            SigAlgo = sigAlgo,
            ManifestHashAlgo = manifestHashAlgo,
            Flags = 0,
            SignerKeyFingerprint = (byte[])signerKeyFingerprint.Clone(),
            SignedAtUnixSeconds = signedAtUnixSeconds,
            SignedEntries = signedEntries,
        };
    }

    /// <summary>Serialised length in bytes.</summary>
    public int ByteLen() =>
        Constants.ManifestPrefixSize + Constants.SignedEntrySize * SignedEntries.Count;

    /// <summary>Serialise to the on-disk byte layout (spec Section 7.1).</summary>
    public byte[] ToBytes()
    {
        var out_ = new byte[ByteLen()];
        Buffer.BlockCopy(Constants.SigMagic, 0, out_, 0, 8);
        LittleEndian.WriteU16(out_, 8, VersionMajor);
        LittleEndian.WriteU16(out_, 10, VersionMinor);
        out_[12] = SigAlgo.Id();
        out_[13] = ManifestHashAlgo.Id();
        LittleEndian.WriteU16(out_, 14, Flags);
        Buffer.BlockCopy(SignerKeyFingerprint, 0, out_, 16, Constants.FingerprintSize);
        LittleEndian.WriteI64(out_, 48, SignedAtUnixSeconds);
        LittleEndian.WriteU32(out_, 56, (uint)SignedEntries.Count);
        for (int i = 0; i < SignedEntries.Count; i++)
        {
            Buffer.BlockCopy(
                SignedEntries[i].ToBytes(), 0,
                out_, Constants.ManifestPrefixSize + i * Constants.SignedEntrySize,
                Constants.SignedEntrySize);
        }
        return out_;
    }

    /// <summary>
    /// Parse from the on-disk byte layout. Validates magic, major version,
    /// algorithm registry membership, hash-algo binding, cryptographic hash
    /// requirement, reserved flags, non-empty signed_count, and per-entry
    /// reserved spans.
    /// </summary>
    public static Manifest FromBytes(byte[] b)
    {
        if (b == null || b.Length < Constants.ManifestPrefixSize)
        {
            throw PcfSigException.MalformedSignaturePartition();
        }
        for (int i = 0; i < 8; i++)
        {
            if (b[i] != Constants.SigMagic[i])
            {
                throw PcfSigException.BadManifestMagic();
            }
        }
        ushort versionMajor = LittleEndian.ReadU16(b, 8);
        ushort versionMinor = LittleEndian.ReadU16(b, 10);
        if (versionMajor != Constants.ProfileVersionMajor)
        {
            throw PcfSigException.UnsupportedMajor(versionMajor);
        }
        var sigAlgo = SigAlgoExtensions.FromId(b[12]);
        byte manifestHashId = b[13];
        var manifestHashAlgo = HashAlgoExtensions.FromId(manifestHashId);
        if (!IsCryptoHash(manifestHashAlgo))
        {
            throw PcfSigException.NonCryptoManifestHash(manifestHashId);
        }
        var required = sigAlgo.RequiredManifestHash();
        if (required.HasValue && required.Value != manifestHashAlgo)
        {
            throw PcfSigException.HashAlgoBindingMismatch();
        }
        ushort flags = LittleEndian.ReadU16(b, 14);
        if (flags != 0)
        {
            throw PcfSigException.NonZeroFlags();
        }
        var signerKeyFingerprint = new byte[Constants.FingerprintSize];
        Buffer.BlockCopy(b, 16, signerKeyFingerprint, 0, Constants.FingerprintSize);
        long signedAtUnixSeconds = LittleEndian.ReadI64(b, 48);
        uint signedCount = LittleEndian.ReadU32(b, 56);
        if (signedCount == 0)
        {
            throw PcfSigException.EmptyManifest();
        }
        int expected = Constants.ManifestPrefixSize + Constants.SignedEntrySize * (int)signedCount;
        if (b.Length < expected)
        {
            throw PcfSigException.MalformedSignaturePartition();
        }
        var entries = new List<SignedEntry>((int)signedCount);
        var seen = new HashSet<string>();
        for (uint i = 0; i < signedCount; i++)
        {
            int off = Constants.ManifestPrefixSize + (int)i * Constants.SignedEntrySize;
            var slice = new byte[Constants.SignedEntrySize];
            Buffer.BlockCopy(b, off, slice, 0, Constants.SignedEntrySize);
            var e = SignedEntry.FromBytes(slice);
            string key = BitConverter.ToString(e.Uid);
            if (!seen.Add(key))
            {
                throw PcfSigException.DuplicateSignedUid();
            }
            entries.Add(e);
        }
        return new Manifest
        {
            VersionMajor = versionMajor,
            VersionMinor = versionMinor,
            SigAlgo = sigAlgo,
            ManifestHashAlgo = manifestHashAlgo,
            Flags = flags,
            SignerKeyFingerprint = signerKeyFingerprint,
            SignedAtUnixSeconds = signedAtUnixSeconds,
            SignedEntries = entries,
        };
    }

    /// <summary>Whether a PCF hash algorithm id is cryptographic (spec Section 9).</summary>
    public static bool IsCryptoHash(HashAlgo a) =>
        a == HashAlgo.Sha256 || a == HashAlgo.Sha512 || a == HashAlgo.Blake3;
}

